File: src/test/java/dev/jlo/gamemodes/arena/ArenaBehaviorTest.java
package dev.jlo.gamemodes.arena;

import org.junit.jupiter.api.Test;

import java.nio.file.Path;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

class ArenaBehaviorTest {
    @Test
    void definitionRejectsUnsafeTemplateAndPlacements() {
        assertThrows(IllegalArgumentException.class, () -> definition("../escape", null));
        assertThrows(IllegalArgumentException.class, () -> definition(null,
                List.of(new Spawn("a", TeamSlot.A, new BlockPosition(1, 1, 1)))));
    }

    @Test
    void regionOverlapIsInclusive() {
        assertTrue(new Region(0, 0, 0, 1, 1, 1)
                .overlaps(new Region(1, 1, 1, 2, 2, 2)));
    }

    @Test
    void generationCleanupCannotRemoveNewerReservation() {
        ArenaFilesystem filesystem = new ArenaFilesystem() {
            @Override
            public boolean exists(Path path) {
                return true;
            }

            @Override
            public void deleteRecursively(Path path) {
            }
        };
        ArenaWorldGateway worlds = new ArenaWorldGateway() {
            @Override
            public void load(Path template, Path instance) {
            }

            @Override
            public void unload(Path instance) {
            }
        };
        ArenaInstanceManager manager = new ArenaInstanceManager(
                Path.of("templates"), Path.of("instances"), filesystem, worlds);
        ArenaDefinition definition = definition(null, null);
        manager.load(definition, new GenerationId("g-2"));
        assertFalse(manager.unload(definition.getId(), new GenerationId("g-1")));
        assertTrue(manager.unload(definition.getId(), new GenerationId("g-2")));
    }

    private static ArenaDefinition definition(String templateWorld, List<Spawn> spawns) {
        if (templateWorld == null) {
            templateWorld = "opr-one";
        }
        if (spawns == null) {
            spawns = List.of(
                    new Spawn("a", TeamSlot.A, new BlockPosition(1, 1, 1)),
                    new Spawn("b", TeamSlot.B, new BlockPosition(9, 1, 9)));
        }
        return new ArenaDefinition(
                "arena",
                ArenaMode.OPR,
                templateWorld,
                new Region(0, 0, 0, 10, 10, 10),
                spawns,
                List.of(
                        new Objective("a", new Region(1, 1, 1, 2, 2, 2)),
                        new Objective("b", new Region(4, 1, 4, 5, 2, 5)),
                        new Objective("c", new Region(7, 1, 7, 8, 2, 8))),
                List.of(
                        new StructurePlacement("gate-a", StructureKind.GATE, new BlockPosition(1, 1, 1)),
                        new StructurePlacement("gate-b", StructureKind.GATE, new BlockPosition(9, 1, 9)),
                        new StructurePlacement("command-a", StructureKind.COMMAND_POST, new BlockPosition(2, 1, 2)),
                        new StructurePlacement("command-b", StructureKind.COMMAND_POST, new BlockPosition(8, 1, 8))),
                List.of(new ResourcePlacement("ore", ResourceKind.ORE, new BlockPosition(3, 1, 3))),
                List.of());
    }
}

File: src/test/kotlin/dev/jlo/gamemodes/arena/ArenaBehaviorTest.kt
package dev.jlo.gamemodes.arena

import kotlin.test.Test
import kotlin.test.assertFailsWith
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class ArenaBehaviorTest {
    @Test
    fun `definition rejects unsafe template and placements`() {
        assertFailsWith<IllegalArgumentException> { definition(templateWorld = "../escape") }
        assertFailsWith<IllegalArgumentException> {
            definition(spawns = listOf(Spawn("a", TeamSlot.A, BlockPosition(1, 1, 1))))
        }
    }

    @Test
    fun `region overlap is inclusive`() {
        assertTrue(Region(0, 0, 0, 1, 1, 1).overlaps(Region(1, 1, 1, 2, 2, 2)))
    }

    @Test
    fun `generation cleanup cannot remove newer reservation`() {
        val fs = object : ArenaFilesystem {
            override fun exists(path: java.nio.file.Path) = true
            override fun deleteRecursively(path: java.nio.file.Path) {}
        }
        val worlds = object : ArenaWorldGateway {
            override fun load(template: java.nio.file.Path, instance: java.nio.file.Path) {}
            override fun unload(instance: java.nio.file.Path) {}
        }
        val manager = ArenaInstanceManager(java.nio.file.Path.of("templates"), java.nio.file.Path.of("instances"), fs, worlds)
        val definition = definition()
        manager.load(definition, GenerationId("g-2"))
        assertFalse(manager.unload(definition.id, GenerationId("g-1")))
        assertTrue(manager.unload(definition.id, GenerationId("g-2")))
    }

    private fun definition(
        templateWorld: String = "opr-one",
        spawns: List<Spawn> = listOf(
            Spawn("a", TeamSlot.A, BlockPosition(1, 1, 1)),
            Spawn("b", TeamSlot.B, BlockPosition(9, 1, 9))
        )
    ) = ArenaDefinition(
        "arena",
        ArenaMode.OPR,
        templateWorld,
        Region(0, 0, 0, 10, 10, 10),
        spawns,
        listOf(
            Objective("a", Region(1, 1, 1, 2, 2, 2)),
            Objective("b", Region(4, 1, 4, 5, 2, 5)),
            Objective("c", Region(7, 1, 7, 8, 2, 8))
        ),
        listOf(
            StructurePlacement("gate-a", StructureKind.GATE, BlockPosition(1, 1, 1)),
            StructurePlacement("gate-b", StructureKind.GATE, BlockPosition(9, 1, 9)),
            StructurePlacement("command-a", StructureKind.COMMAND_POST, BlockPosition(2, 1, 2)),
            StructurePlacement("command-b", StructureKind.COMMAND_POST, BlockPosition(8, 1, 8))
        ),
        listOf(ResourcePlacement("ore", ResourceKind.ORE, BlockPosition(3, 1, 3))),
        emptyList()
    )
}
